diff --git a/arch/arm/kernel/perf_event_cpu.c b/arch/arm/kernel/perf_event_cpu.c
index e0665b8..296e531 100644
--- a/arch/arm/kernel/perf_event_cpu.c
+++ b/arch/arm/kernel/perf_event_cpu.c
@@ -348,6 +348,13 @@ static struct platform_driver cpu_pmu_driver = {
 	.id_table	= cpu_pmu_plat_device_ids,
 };
 
+#ifdef CONFIG_PMCTRACK
+static int __init register_pmu_driver(void)
+{
+        pr_cont("PMCTRACK enabled: only software events available in perf.\n");
+        return 0;
+}   
+#else
 static int __init register_pmu_driver(void)
 {
 	int err;
@@ -370,4 +377,5 @@ static int __init register_pmu_driver(void)
 
 	return err;
 }
+#endif
 device_initcall(register_pmu_driver);
diff --git a/arch/arm64/kernel/perf_event.c b/arch/arm64/kernel/perf_event.c
index dfcd8fa..48f2d70 100644
--- a/arch/arm64/kernel/perf_event.c
+++ b/arch/arm64/kernel/perf_event.c
@@ -1320,6 +1320,13 @@ static void __init cpu_pmu_init(struct arm_pmu *armpmu)
 	armpmu->get_hw_events = armpmu_get_cpu_events;
 }
 
+#ifdef CONFIG_PMCTRACK
+static int __init init_hw_perf_events(void)
+{
+        pr_cont("PMCTRACK enabled: only software events available in perf.\n");
+        return 0;
+}
+#else
 static int __init init_hw_perf_events(void)
 {
 	u64 dfr = read_cpuid(ID_AA64DFR0_EL1);
@@ -1341,6 +1348,7 @@ static int __init init_hw_perf_events(void)
 
 	return 0;
 }
+#endif
 early_initcall(init_hw_perf_events);
 
 /*
diff --git a/include/linux/pmctrack.h b/include/linux/pmctrack.h
new file mode 100644
index 0000000..523b0ea
--- /dev/null
+++ b/include/linux/pmctrack.h
@@ -0,0 +1,77 @@
+/*
+ *  include/linux/pmctrack.h
+ *
+ *  Copyright (c) 2015 Juan Carlos Saez <jcsaezal@ucm.es>
+ * 
+ *  This code is licensed under the GNU GPL v2.
+ */
+/* 
+ *   Written by Juan Carlos Saez with help from
+ * 	 Guillermo Martinez Fernandez, 
+ *	 Sergio Sanchez Gordo and Sofia Dronda Merino 
+ * 
+ */
+
+#ifndef PMCTRACK_H
+#define PMCTRACK_H
+#include <linux/types.h>
+#include <linux/module.h>
+#include <linux/device.h>
+
+/*** Interface to interact with monitoring modules ***/
+
+/* Predefined high-level metrics */
+typedef enum {
+	MC_SPEEDUP_FACTOR,
+	MC_INSTR_PER_CYCLE,
+	MC_LLC_MISSES_PER_MINSTR,
+	MC_LLC_REQUESTS_PER_KINSTR
+} mc_metric_key_t;
+
+int pmcs_get_current_metric_value(struct task_struct* task, int key, uint64_t* value);
+/******************************************************************/
+
+/* Interface for PMCTrack kernel module */
+typedef struct __pmc_ops {
+	int 	(*pmcs_alloc_per_thread_data)(unsigned long,struct task_struct*);
+	void	(*pmcs_save_callback)(void*, int);
+	void	(*pmcs_restore_callback)(void*, int);
+	void 	(*pmcs_tbs_tick)(void*, int);
+	void	(*pmcs_exec_thread)(struct task_struct*);
+	void	(*pmcs_free_per_thread_data)(struct task_struct*);
+	void	(*pmcs_exit_thread)(struct task_struct*);
+	int	(*pmcs_get_current_metric_value)(struct task_struct* task, int key, uint64_t* value);
+} pmc_ops_t;
+
+
+/* Register/Unregister implementation */
+int register_pmc_module(pmc_ops_t* pmc_ops_module, struct module* module);
+int unregister_pmc_module(pmc_ops_t* pmc_ops_module, struct module* module);
+
+/* PMCTrack kernel API */
+int pmcs_alloc_per_thread_data(unsigned long clone_flags, struct task_struct *p);
+void pmcs_save_callback(void* prof, int cpu);
+void pmcs_restore_callback(void* prof, int cpu);
+void pmcs_tbs_tick(void* prof, int cpu);
+void pmcs_exec_thread(struct task_struct* tsk);
+void pmcs_free_per_thread_data(struct task_struct* tsk);
+void pmcs_exit_thread(struct task_struct* tsk);
+
+
+/*** Hwmon-PMCTrack bridge code ***/
+
+/* Export/Unexport devices */
+int pmctrack_hwmon_export_device(const char* name, int id, struct device* dev);
+void pmctrack_hwmon_unexport_device(int id);
+
+/* Retrieve device and increase ref counter */
+struct device* pmctrack_hwmon_get_device(const char* name);
+
+/* Decrease ref count */
+void pmctrack_hwmon_put_device(struct device* dev);
+
+/* Show registered devices */
+void pmctrack_hwmon_list_devices(int max_devices, const char* strs[], int* nr_devices);
+
+
+#endif
diff --git a/include/linux/sched.h b/include/linux/sched.h
index 5229df9..6c4a24e 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1448,6 +1448,10 @@ struct task_struct {
 	unsigned int	sequential_io;
 	unsigned int	sequential_io_avg;
 #endif
+#ifdef CONFIG_PMCTRACK
+        void *pmc;                      /* Per-thread PMC-specific data */
+        unsigned char prof_enabled;     /* This field must be one for the profiler to be active in the current task */
+#endif
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
diff --git a/init/Kconfig b/init/Kconfig
index 79b356f..5dbc043 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1507,6 +1507,15 @@ config DEBUG_PERF_USE_VMALLOC
 
 	 Say N if unsure.
 
+config PMCTRACK
+        bool "PMCTrack performance monitoring tool with in-kernel interface"
+        default y
+        help
+          PMCTrack performance monitoring tool with in-kernel interface.
+          Warning: Activating this tool automatically disables HW events in
+          the perf events subsystem on ARM and x86 platforms (including 32 and
+          64 bit variants).
+
 endmenu
 
 config VM_EVENT_COUNTERS
diff --git a/kernel/Makefile b/kernel/Makefile
index 271fd31..1ec5661 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -26,6 +26,7 @@ obj-y += sched/
 obj-y += power/
 obj-y += cpu/
 
+obj-$(CONFIG_PMCTRACK) += pmctrack.o
 obj-$(CONFIG_CHECKPOINT_RESTORE) += kcmp.o
 obj-$(CONFIG_FREEZER) += freezer.o
 obj-$(CONFIG_PROFILING) += profile.o
diff --git a/kernel/exit.c b/kernel/exit.c
index 33fde71..31a2bcf 100644
--- a/kernel/exit.c
+++ b/kernel/exit.c
@@ -53,6 +53,7 @@
 #include <linux/oom.h>
 #include <linux/writeback.h>
 #include <linux/shm.h>
+#include <linux/pmctrack.h>
 
 #include <asm/uaccess.h>
 #include <asm/unistd.h>
@@ -817,6 +818,9 @@ void do_exit(long code)
 	 */
 	perf_event_exit_task(tsk);
 
+#ifdef CONFIG_PMCTRACK
+        pmcs_exit_thread(tsk);
+#endif
 	cgroup_exit(tsk, 1);
 
 	module_put(task_thread_info(tsk)->exec_domain->module);
diff --git a/kernel/fork.c b/kernel/fork.c
index b066395..55d7015 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -80,6 +80,7 @@
 #include <asm/tlbflush.h>
 
 #include <trace/events/sched.h>
+#include <linux/pmctrack.h>
 
 #define CREATE_TRACE_POINTS
 #include <trace/events/task.h>
@@ -210,6 +211,9 @@ static void account_kernel_stack(struct thread_info *ti, int account)
 
 void free_task(struct task_struct *tsk)
 {
+#ifdef CONFIG_PMCTRACK
+        pmcs_free_per_thread_data(tsk);
+#endif
 	account_kernel_stack(tsk->stack, -1);
 	arch_release_thread_info(tsk->stack);
 	free_thread_info(tsk->stack);
@@ -1374,6 +1378,12 @@ static struct task_struct *copy_process(unsigned long clone_flags,
 	p->sequential_io_avg	= 0;
 #endif
 
+#ifdef CONFIG_PMCTRACK
+        p->prof_enabled = 0;    /* The profiler will be disabled by default */
+        p->pmc = NULL;
+        if (pmcs_alloc_per_thread_data(clone_flags,p))
+                goto bad_fork_cleanup_policy;
+#endif  
 	/* Perform scheduler related setup. Assign this task to a CPU. */
 	sched_fork(p);
 
diff --git a/kernel/futex.c b/kernel/futex.c
index ad971d0..8865f53 100644
--- a/kernel/futex.c
+++ b/kernel/futex.c
@@ -194,6 +194,8 @@ static void get_futex_key_refs(union futex_key *key)
 	case FUT_OFF_MMSHARED:
 		atomic_inc(&key->private.mm->mm_count);
 		break;
+	default:
+		smp_mb(); /* explicit MB (B) */
 	}
 }
 
diff --git a/kernel/irq/manage.c b/kernel/irq/manage.c
index a79d267..7afbb02 100644
--- a/kernel/irq/manage.c
+++ b/kernel/irq/manage.c
@@ -202,6 +202,9 @@ int __irq_set_affinity(unsigned int irq, const struct cpumask *mask, bool force)
 	raw_spin_unlock_irqrestore(&desc->lock, flags);
 	return ret;
 }
+#ifdef CONFIG_PMCTRACK
+EXPORT_SYMBOL_GPL(__irq_set_affinity);
+#endif
 
 int irq_set_affinity_hint(unsigned int irq, const struct cpumask *m)
 {
diff --git a/kernel/pmctrack.c b/kernel/pmctrack.c
new file mode 100644
index 0000000..80549cf
--- /dev/null
+++ b/kernel/pmctrack.c
@@ -0,0 +1,389 @@
+/*
+ *  kernel/pmctrack.c
+ *
+ *  Copyright (c) 2015 Juan Carlos Saez <jcsaezal@ucm.es>
+ * 
+ *  This code is licensed under the GNU GPL v2.
+ */
+/* 
+ *   Written by Juan Carlos Saez with help from
+ * 	 Guillermo Martinez Fernandez, 
+ *	 Sergio Sanchez Gordo and Sofia Dronda Merino 
+ * 
+ */
+
+#include <linux/pmctrack.h>
+#include <linux/module.h>
+#include <linux/list.h>
+#include <linux/device.h>
+#include <linux/err.h>
+#include <linux/slab.h>
+#include <linux/rwlock.h>
+#include <linux/rcupdate.h>
+
+static pmc_ops_t* pmc_ops_mod=NULL; /* No implementation is registered by default */
+static struct module* implementer=NULL;
+DEFINE_RWLOCK(pmc_ops_lock);
+
+/* 
+ * PMCTrack's kernel module invokes this function to register 
+ * an implementation of the pmc_ops_t interface
+ */
+int register_pmc_module(pmc_ops_t* pmc_ops_module, struct module* module)
+{
+	int ret=0;
+	unsigned long flags;
+
+	write_lock_irqsave(&pmc_ops_lock,flags);
+	
+	/* Module has been installed already */
+	if (implementer!=NULL) {
+		ret=-EPERM;
+	}
+	else {
+		implementer = module;
+		rcu_assign_pointer(pmc_ops_mod,pmc_ops_module);
+	}
+
+	write_unlock_irqrestore(&pmc_ops_lock,flags);
+	return ret;
+}
+
+
+/* PMCTrack's kernel module invokes this function when unloaded */
+int unregister_pmc_module(pmc_ops_t* pmc_ops_module, struct module* module)
+{
+	int ret=0;
+	unsigned long flags;
+
+	write_lock_irqsave(&pmc_ops_lock,flags);
+
+	if(implementer!=module){
+		ret=-EPERM;
+	}
+	else { 
+		implementer=NULL;
+		rcu_assign_pointer(pmc_ops_mod,NULL);
+	}
+	write_unlock_irqrestore(&pmc_ops_lock,flags);
+
+	/* 
+	 * If the operation succeeded wait for all readers to complete. 
+	 * Since synchronize_rcu() may block, this has to be done 
+	 * without the spin lock held
+	 */
+	if (ret==0)
+		synchronize_rcu();
+
+	return ret;
+}
+
+EXPORT_SYMBOL(register_pmc_module);
+EXPORT_SYMBOL(unregister_pmc_module);
+
+
+/*
+ * Wrapper function for the various pmc_ops_t operations 
+ */
+
+/* Invoked when forking a process/thread */ 
+int pmcs_alloc_per_thread_data(unsigned long clone_flags, struct task_struct *p)
+{
+	int ret=0;
+	pmc_ops_t* pmc_ops= NULL;
+	unsigned long flags;
+
+	read_lock_irqsave(&pmc_ops_lock,flags);
+
+	/* 
+	 * If there is no implementer module or it's being removed
+	 * from the kernel, return immediately.
+	 */
+	if (!implementer || !try_module_get(implementer)){
+		read_unlock_irqrestore(&pmc_ops_lock,flags);
+		return 0;	
+	}
+
+	read_unlock_irqrestore(&pmc_ops_lock,flags);
+	
+	/* Now it's safe to dereference pmc_ops_mod */
+	pmc_ops=pmc_ops_mod;
+
+	/* Invoke the allocation operation (may block) */
+	if(pmc_ops!=NULL && pmc_ops->pmcs_alloc_per_thread_data!=NULL)
+		ret=pmc_ops->pmcs_alloc_per_thread_data(clone_flags,p);
+
+	/* Allow the module to be removed now */
+	module_put(implementer);
+
+	return ret;
+}
+
+/* Invoked when a context switch out takes place */ 
+void pmcs_save_callback(void *prof, int cpu)
+{
+	pmc_ops_t* pmc_ops= NULL;
+
+	if (!implementer)
+		return;
+
+	rcu_read_lock();
+
+	pmc_ops=rcu_dereference(pmc_ops_mod);
+
+	if(pmc_ops!=NULL && pmc_ops->pmcs_save_callback!=NULL)
+		pmc_ops->pmcs_save_callback(prof, cpu);
+
+	rcu_read_unlock();
+}
+
+/* Invoked when a context switch in takes place */ 
+void pmcs_restore_callback(void *prof, int cpu)
+{
+	pmc_ops_t* pmc_ops= NULL;
+
+	if (!implementer)
+		return;
+
+	rcu_read_lock();
+
+	pmc_ops=rcu_dereference(pmc_ops_mod);
+
+	if(pmc_ops!=NULL && pmc_ops->pmcs_restore_callback!=NULL)
+		pmc_ops->pmcs_restore_callback(prof, cpu);
+
+	rcu_read_unlock();
+}
+
+/* Invoked from scheduler_tick() */ 
+void pmcs_tbs_tick(void *prof, int cpu)
+{
+	pmc_ops_t* pmc_ops= NULL;
+
+	if (!implementer)
+		return;
+
+	rcu_read_lock();
+
+	pmc_ops=rcu_dereference(pmc_ops_mod);
+
+	if(pmc_ops!=NULL && pmc_ops->pmcs_tbs_tick!=NULL)
+		pmc_ops->pmcs_tbs_tick(prof, cpu);
+
+	rcu_read_unlock();
+}
+
+/* Invoked when a process calls exec() */ 
+void pmcs_exec_thread(struct task_struct* tsk)
+{
+	pmc_ops_t* pmc_ops= NULL;
+
+	if (!implementer)
+		return;
+
+	rcu_read_lock();
+
+	pmc_ops=rcu_dereference(pmc_ops_mod);
+
+	if(pmc_ops!=NULL && pmc_ops->pmcs_exec_thread!=NULL)
+		pmc_ops->pmcs_exec_thread(tsk);
+
+	rcu_read_unlock();
+}
+
+/* Invoked when the kernel frees up the process descriptor */ 
+void pmcs_free_per_thread_data(struct task_struct* tsk)
+{
+	pmc_ops_t* pmc_ops= NULL;
+	unsigned long flags;
+
+	read_lock_irqsave(&pmc_ops_lock,flags);
+
+	/* 
+	 * If there is no implementer module or it's being removed
+	 * from the kernel, return immediately.
+	 */
+	if (!implementer || !try_module_get(implementer)){
+		read_unlock_irqrestore(&pmc_ops_lock,flags);
+		return;
+	}
+
+	read_unlock_irqrestore(&pmc_ops_lock,flags);
+	
+	/* Now it's safe to dereference pmc_ops_mod */
+	pmc_ops=pmc_ops_mod;
+	
+	if(pmc_ops!=NULL && pmc_ops->pmcs_free_per_thread_data!=NULL)
+		pmc_ops->pmcs_free_per_thread_data(tsk);
+
+	/* Allow the module to be removed now */
+        module_put(implementer);
+}
+
+/* Invoked when a process exits */ 
+void pmcs_exit_thread(struct task_struct* tsk)
+{
+	pmc_ops_t* pmc_ops= NULL;
+	unsigned long flags;
+
+	read_lock_irqsave(&pmc_ops_lock,flags);
+
+	/* 
+	 * If there is no implementer module or it's being removed
+	 * from the kernel, return immediately.
+	 */
+	if (!implementer || !try_module_get(implementer)){
+		read_unlock_irqrestore(&pmc_ops_lock,flags);
+		return;	
+	}
+
+	read_unlock_irqrestore(&pmc_ops_lock,flags);
+	
+	/* Now it's safe to dereference pmc_ops_mod */
+	pmc_ops=pmc_ops_mod;
+
+	if(pmc_ops!=NULL && pmc_ops->pmcs_exit_thread!=NULL)
+		pmc_ops->pmcs_exit_thread(tsk);
+
+	/* Allow the module to be removed now */
+        module_put(implementer);
+}
+
+/* 
+ * Invoked from the code of experimental scheduling classes that leverage per-thread performance
+ * counter data when making scheduling decisions. 
+ * The source code of these scheduling classes is not provided along with this patch, though.  
+ */ 
+int pmcs_get_current_metric_value(struct task_struct* task, int key, uint64_t* value)
+{
+	int ret=-1;
+	pmc_ops_t* pmc_ops= NULL;
+
+	if (!implementer)
+		return ret;
+
+	rcu_read_lock();
+
+	pmc_ops=rcu_dereference(pmc_ops_mod);
+
+	if(pmc_ops!=NULL && pmc_ops->pmcs_get_current_metric_value!=NULL)
+		ret=pmc_ops->pmcs_get_current_metric_value(task,key,value);
+
+	rcu_read_unlock();
+
+	return ret;
+}
+
+/* Hwmon-PMCTRack bridge code */
+typedef struct {
+	const char* name;
+	int id;
+	struct device* dev;
+	struct list_head links;
+} pmctrack_hwmon_node_t;
+
+
+LIST_HEAD(pmctrack_hwmon_devices);
+
+/* Spinlock to protect list of devices */
+DEFINE_SPINLOCK(pmctrack_hwmon_sp);
+
+/* Export/Unexport Functions */
+int pmctrack_hwmon_export_device(const char* name, int id, struct device* dev)
+{
+	pmctrack_hwmon_node_t* node;
+
+	node = kzalloc(sizeof(pmctrack_hwmon_node_t), GFP_KERNEL);
+	if (node == NULL)
+		return -ENOMEM;
+
+	node->name=name;
+	node->id=id;
+	node->dev=dev;
+
+	spin_lock(&pmctrack_hwmon_sp);
+	list_add_tail(&node->links,&pmctrack_hwmon_devices);
+	spin_unlock(&pmctrack_hwmon_sp);
+
+	return 0;
+}
+EXPORT_SYMBOL(pmctrack_hwmon_export_device);
+
+void pmctrack_hwmon_unexport_device(int id)
+{
+	pmctrack_hwmon_node_t* item=NULL;
+	struct list_head* cur_node=NULL;
+
+	spin_lock(&pmctrack_hwmon_sp);
+
+	list_for_each(cur_node, &pmctrack_hwmon_devices) {
+		item = list_entry(cur_node, pmctrack_hwmon_node_t, links);
+		if (item->id==id) {
+			list_del(cur_node);
+			kfree(item);
+			goto out_unexport;
+		}
+	}
+out_unexport:
+	spin_unlock(&pmctrack_hwmon_sp);
+}
+EXPORT_SYMBOL(pmctrack_hwmon_unexport_device);
+
+/* Retrieve device and increase ref coutner */
+struct device* pmctrack_hwmon_get_device(const char* name)
+{
+	pmctrack_hwmon_node_t* item=NULL;
+	struct list_head* cur_node=NULL;
+	struct device* ret=NULL;
+
+	spin_lock(&pmctrack_hwmon_sp);
+
+	list_for_each(cur_node, &pmctrack_hwmon_devices) {
+		item = list_entry(cur_node, pmctrack_hwmon_node_t, links);
+		if (strcmp(item->name,name)==0) {
+			get_device(item->dev);
+			ret=item->dev;
+			goto out_get_device;
+		}
+	}
+out_get_device:
+	spin_unlock(&pmctrack_hwmon_sp);
+	return ret;
+}
+EXPORT_SYMBOL(pmctrack_hwmon_get_device);
+
+
+/* Decrease ref count */
+void pmctrack_hwmon_put_device(struct device* dev)
+{
+	put_device(dev);
+}
+EXPORT_SYMBOL(pmctrack_hwmon_put_device);
+
+/* Show registered devices */
+void pmctrack_hwmon_list_devices(int max_devices, const char* strs[], int* nr_devices)
+{
+	pmctrack_hwmon_node_t* item=NULL;
+	struct list_head* cur_node=NULL;
+	int cnt=0;
+
+	if (!max_devices) {
+		(*nr_devices)=0;
+		return;
+	}
+
+	spin_lock(&pmctrack_hwmon_sp);
+
+	list_for_each(cur_node, &pmctrack_hwmon_devices) {
+		item = list_entry(cur_node, pmctrack_hwmon_node_t, links);
+		strs[cnt++]=item->name;
+		if (cnt>=max_devices)
+			goto out_list_devices;
+	}
+out_list_devices:
+	(*nr_devices)=cnt;
+	spin_unlock(&pmctrack_hwmon_sp);
+}
+
+EXPORT_SYMBOL(pmctrack_hwmon_list_devices);
+
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index ea4e780..8dfb9dc 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -73,6 +73,7 @@
 #include <linux/init_task.h>
 #include <linux/binfmts.h>
 #include <linux/context_tracking.h>
+#include <linux/pmctrack.h>
 
 #include <asm/switch_to.h>
 #include <asm/tlb.h>
@@ -1706,7 +1707,6 @@ void sched_fork(struct task_struct *p)
 	 * event cannot wake it up and insert it on the runqueue either.
 	 */
 	p->state = TASK_RUNNING;
-
 	/*
 	 * Make sure we do not leak PI boosting priority to the child.
 	 */
@@ -1921,6 +1921,9 @@ static void finish_task_switch(struct rq *rq, struct task_struct *prev)
 	vtime_task_switch(prev);
 	finish_arch_switch(prev);
 	perf_event_task_sched_in(prev, current);
+#ifdef CONFIG_PMCTRACK
+        pmcs_restore_callback(current->pmc, smp_processor_id());
+#endif
 	finish_lock_switch(rq, prev);
 	finish_arch_post_lock_switch();
 
@@ -2679,6 +2682,9 @@ void sched_exec(void)
 	unsigned long flags;
 	int dest_cpu;
 
+#ifdef CONFIG_PMCTRACK
+        pmcs_exec_thread(p);
+#endif
 	raw_spin_lock_irqsave(&p->pi_lock, flags);
 	dest_cpu = p->sched_class->select_task_rq(p, SD_BALANCE_EXEC, 0);
 	if (dest_cpu == smp_processor_id())
@@ -2764,6 +2770,9 @@ void scheduler_tick(void)
 	struct rq *rq = cpu_rq(cpu);
 	struct task_struct *curr = rq->curr;
 
+#ifdef CONFIG_PMCTRACK
+        pmcs_tbs_tick(curr->pmc, cpu);
+#endif
 	sched_clock_tick();
 
 	raw_spin_lock(&rq->lock);
@@ -3042,6 +3051,9 @@ need_resched:
 		rq->curr = next;
 		++*switch_count;
 
+#ifdef CONFIG_PMCTRACK
+                pmcs_save_callback(prev->pmc, cpu);
+#endif
 		context_switch(rq, prev, next); /* unlocks the rq */
 		/*
 		 * The context switch have flipped the stack from under us
@@ -3843,10 +3855,17 @@ struct task_struct *idle_task(int cpu)
  * find_process_by_pid - find a process with a matching PID value.
  * @pid: the pid in question.
  */
+#ifdef CONFIG_PMCTRACK
+struct task_struct *find_process_by_pid(pid_t pid)
+#else
 static struct task_struct *find_process_by_pid(pid_t pid)
+#endif
 {
 	return pid ? find_task_by_vpid(pid) : current;
 }
+#ifdef CONFIG_PMCTRACK
+EXPORT_SYMBOL_GPL(find_process_by_pid);
+#endif
 
 extern struct cpumask hmp_slow_cpu_mask;
 
diff --git a/scripts/package/builddeb b/scripts/package/builddeb
index 3001ec5..0d64196 100644
--- a/scripts/package/builddeb
+++ b/scripts/package/builddeb
@@ -35,15 +35,17 @@ create_package() {
 	sparc*)
 		debarch=sparc ;;
 	s390*)
-		debarch=s390 ;;
+		debarch=s390$(grep -q CONFIG_64BIT=y $KCONFIG_CONFIG && echo x || true) ;;
 	ppc*)
 		debarch=powerpc ;;
 	parisc*)
 		debarch=hppa ;;
 	mips*)
-		debarch=mips$(grep -q CPU_LITTLE_ENDIAN=y .config && echo el) ;;
+		debarch=mips$(grep -q CPU_LITTLE_ENDIAN=y $KCONFIG_CONFIG && echo el || true) ;;
+	arm64)
+		debarch=arm64 ;;
 	arm*)
-		debarch=arm$(grep -q CONFIG_AEABI=y .config && echo el) ;;
+		debarch=arm$(grep -q CONFIG_AEABI=y $KCONFIG_CONFIG && echo el || true) ;;
 	*)
 		echo "" >&2
 		echo "** ** **  WARNING  ** ** **" >&2
@@ -78,17 +80,38 @@ tmpdir="$objtree/debian/tmp"
 fwdir="$objtree/debian/fwtmp"
 kernel_headers_dir="$objtree/debian/hdrtmp"
 libc_headers_dir="$objtree/debian/headertmp"
+dbg_dir="$objtree/debian/dbgtmp"
 packagename=linux-image-$version
-fwpackagename=linux-firmware-image
+fwpackagename=linux-firmware-image-$version
 kernel_headers_packagename=linux-headers-$version
 libc_headers_packagename=linux-libc-dev
+dbg_packagename=$packagename-dbg
 
 if [ "$ARCH" = "um" ] ; then
 	packagename=user-mode-linux-$version
 fi
 
+# Not all arches have the same installed path in debian
+# XXX: have each arch Makefile export a variable of the canonical image install
+# path instead
+case $ARCH in
+um)
+	installed_image_path="usr/bin/linux-$version"
+	;;
+parisc|mips|powerpc)
+	installed_image_path="boot/vmlinux-$version"
+	;;
+arm64)
+	installed_image_path="boot/Image-$version"
+	;;
+*)
+	installed_image_path="boot/vmlinuz-$version"
+esac
+
+BUILD_DEBUG="$(grep -s '^CONFIG_DEBUG_INFO=y' $KCONFIG_CONFIG || true)"
+
 # Setup the directory structure
-rm -rf "$tmpdir" "$fwdir" "$kernel_headers_dir" "$libc_headers_dir"
+rm -rf "$tmpdir" "$fwdir" "$kernel_headers_dir" "$libc_headers_dir" "$dbg_dir"
 mkdir -m 755 -p "$tmpdir/DEBIAN"
 mkdir -p  "$tmpdir/lib" "$tmpdir/boot" "$tmpdir/usr/share/doc/$packagename"
 mkdir -m 755 -p "$fwdir/DEBIAN"
@@ -101,26 +124,29 @@ mkdir -p "$kernel_headers_dir/lib/modules/$version/"
 if [ "$ARCH" = "um" ] ; then
 	mkdir -p "$tmpdir/usr/lib/uml/modules/$version" "$tmpdir/usr/bin"
 fi
+if [ -n "$BUILD_DEBUG" ] ; then
+	mkdir -p "$dbg_dir/usr/share/doc/$dbg_packagename"
+	mkdir -m 755 -p "$dbg_dir/DEBIAN"
+fi
 
 # Build and install the kernel
 if [ "$ARCH" = "um" ] ; then
 	$MAKE linux
 	cp System.map "$tmpdir/usr/lib/uml/modules/$version/System.map"
-	cp .config "$tmpdir/usr/share/doc/$packagename/config"
+	cp $KCONFIG_CONFIG "$tmpdir/usr/share/doc/$packagename/config"
 	gzip "$tmpdir/usr/share/doc/$packagename/config"
-	cp $KBUILD_IMAGE "$tmpdir/usr/bin/linux-$version"
-else 
+else
 	cp System.map "$tmpdir/boot/System.map-$version"
-	cp .config "$tmpdir/boot/config-$version"
-	# Not all arches include the boot path in KBUILD_IMAGE
-	if [ -e $KBUILD_IMAGE ]; then
-		cp $KBUILD_IMAGE "$tmpdir/boot/vmlinuz-$version"
-	else
-		cp arch/$ARCH/boot/$KBUILD_IMAGE "$tmpdir/boot/vmlinuz-$version"
-	fi
+	cp $KCONFIG_CONFIG "$tmpdir/boot/config-$version"
+fi
+# Not all arches include the boot path in KBUILD_IMAGE
+if [ -e $KBUILD_IMAGE ]; then
+	cp $KBUILD_IMAGE "$tmpdir/$installed_image_path"
+else
+	cp arch/$ARCH/boot/$KBUILD_IMAGE "$tmpdir/$installed_image_path"
 fi
 
-if grep -q '^CONFIG_MODULES=y' .config ; then
+if grep -q '^CONFIG_MODULES=y' $KCONFIG_CONFIG ; then
 	INSTALL_MOD_PATH="$tmpdir" $MAKE KBUILD_SRC= modules_install
 	rm -f "$tmpdir/lib/modules/$version/build"
 	rm -f "$tmpdir/lib/modules/$version/source"
@@ -128,6 +154,20 @@ if grep -q '^CONFIG_MODULES=y' .config ; then
 		mv "$tmpdir/lib/modules/$version"/* "$tmpdir/usr/lib/uml/modules/$version/"
 		rmdir "$tmpdir/lib/modules/$version"
 	fi
+	if [ -n "$BUILD_DEBUG" ] ; then
+		(
+			cd $tmpdir
+			for module in $(find lib/modules/ -name *.ko); do
+				mkdir -p $(dirname $dbg_dir/usr/lib/debug/$module)
+				# only keep debug symbols in the debug file
+				$OBJCOPY --only-keep-debug $module $dbg_dir/usr/lib/debug/$module
+				# strip original module from debug symbols
+				$OBJCOPY --strip-debug $module
+				# then add a link to those
+				$OBJCOPY --add-gnu-debuglink=$dbg_dir/usr/lib/debug/$module $module
+			done
+		)
+	fi
 fi
 
 if [ "$ARCH" != "um" ]; then
@@ -137,8 +177,15 @@ fi
 
 # Install the maintainer scripts
 # Note: hook scripts under /etc/kernel are also executed by official Debian
-# kernel packages, as well as kernel packages built using make-kpkg
+# kernel packages, as well as kernel packages built using make-kpkg.
+# make-kpkg sets $INITRD to indicate whether an initramfs is wanted, and
+# so do we; recent versions of dracut and initramfs-tools will obey this.
 debhookdir=${KDEB_HOOKDIR:-/etc/kernel}
+if grep -q '^CONFIG_BLK_DEV_INITRD=y' $KCONFIG_CONFIG; then
+	want_initrd=Yes
+else
+	want_initrd=No
+fi
 for script in postinst postrm preinst prerm ; do
 	mkdir -p "$tmpdir$debhookdir/$script.d"
 	cat <<EOF > "$tmpdir/DEBIAN/$script"
@@ -149,7 +196,10 @@ set -e
 # Pass maintainer script parameters to hook scripts
 export DEB_MAINT_PARAMS="\$*"
 
-test -d $debhookdir/$script.d && run-parts --arg="$version" $debhookdir/$script.d
+# Tell initramfs builder whether it's wanted
+export INITRD=$want_initrd
+
+test -d $debhookdir/$script.d && run-parts --arg="$version" --arg="/$installed_image_path" $debhookdir/$script.d
 exit 0
 EOF
 	chmod 755 "$tmpdir/DEBIAN/$script"
@@ -242,14 +292,17 @@ EOF
 
 fi
 
-# Build header package
-(cd $srctree; find . -name Makefile\* -o -name Kconfig\* -o -name \*.pl > "$objtree/debian/hdrsrcfiles")
-(cd $srctree; find arch/$SRCARCH/include include scripts -type f >> "$objtree/debian/hdrsrcfiles")
-(cd $objtree; find arch/$SRCARCH/include .config Module.symvers include scripts -type f >> "$objtree/debian/hdrobjfiles")
+# Build kernel header package
+(cd $srctree; find . -name Makefile\* -o -name Kconfig\* -o -name \*.pl) > "$objtree/debian/hdrsrcfiles"
+(cd $srctree; find arch/$SRCARCH/include include scripts -type f) >> "$objtree/debian/hdrsrcfiles"
+(cd $srctree; find arch/$SRCARCH -name module.lds -o -name Kbuild.platforms -o -name Platform) >> "$objtree/debian/hdrsrcfiles"
+(cd $srctree; find $(find arch/$SRCARCH -name include -o -name scripts -type d) -type f) >> "$objtree/debian/hdrsrcfiles"
+(cd $objtree; find arch/$SRCARCH/include Module.symvers include scripts -type f) >> "$objtree/debian/hdrobjfiles"
 destdir=$kernel_headers_dir/usr/src/linux-headers-$version
 mkdir -p "$destdir"
-(cd $srctree; tar -c -f - -T "$objtree/debian/hdrsrcfiles") | (cd $destdir; tar -xf -)
-(cd $objtree; tar -c -f - -T "$objtree/debian/hdrobjfiles") | (cd $destdir; tar -xf -)
+(cd $srctree; tar -c -f - -T -) < "$objtree/debian/hdrsrcfiles" | (cd $destdir; tar -xf -)
+(cd $objtree; tar -c -f - -T -) < "$objtree/debian/hdrobjfiles" | (cd $destdir; tar -xf -)
+(cd $objtree; cp $KCONFIG_CONFIG $destdir/.config) # copy .config manually to be where it's expected to be
 ln -sf "/usr/src/linux-headers-$version" "$kernel_headers_dir/lib/modules/$version/build"
 rm -f "$objtree/debian/hdrsrcfiles" "$objtree/debian/hdrobjfiles"
 
@@ -280,6 +333,17 @@ EOF
 	create_package "$fwpackagename" "$fwdir"
 fi
 
+# Copy device tree files if generated
+stat arch/$ARCH/boot/dts/*.dtb && {
+	mkdir -p "$tmpdir/lib/firmware/$version/device-tree"
+	cp arch/$ARCH/boot/dts/*.dtb "$tmpdir/lib/firmware/$version/device-tree"
+}
+
+# Copy boot wrapper files if built
+if [ "$ARCH" = "arm64" ] ; then
+	stat arch/arm64/boot/*.axf && cp arch/arm64/boot/*.axf "$tmpdir/boot"
+fi
+
 cat <<EOF >> debian/control
 
 Package: $libc_headers_packagename
@@ -298,4 +362,30 @@ fi
 
 create_package "$packagename" "$tmpdir"
 
+if [ -n "$BUILD_DEBUG" ] ; then
+	# Build debug package
+	# Different tools want the image in different locations
+	# perf
+	mkdir -p $dbg_dir/usr/lib/debug/lib/modules/$version/
+	cp vmlinux $dbg_dir/usr/lib/debug/lib/modules/$version/
+	# systemtap
+	mkdir -p $dbg_dir/usr/lib/debug/boot/
+	ln -s ../lib/modules/$version/vmlinux $dbg_dir/usr/lib/debug/boot/vmlinux-$version
+	# kdump-tools
+	ln -s lib/modules/$version/vmlinux $dbg_dir/usr/lib/debug/vmlinux-$version
+
+	cat <<EOF >> debian/control
+
+Package: $dbg_packagename
+Section: debug
+Provides: linux-debug, linux-debug-$version
+Architecture: any
+Description: Linux kernel debugging symbols for $version
+ This package will come in handy if you need to debug the kernel. It provides
+ all the necessary debug symbols for the kernel and its modules.
+EOF
+
+	create_package "$dbg_packagename" "$dbg_dir"
+fi
+
 exit 0
